### DISCUSSION TITLE
Is it possible to call SingleFile from an extension?

### DISCUSSION THREAD

**Discussion by silarri** [gildas-lormeau/SingleFile#1226] [Q&A]
I'm trying to call the SingleFile extension from a custom button that performs several actions that are typical in my workflow. My code in the background.js file is as follows:

```
  const targetExtensionID = "{531906d3-e22f-4a6c-a102-8057b88a1a63}";
  await browser.runtime.sendMessage(targetExtensionID, { action: "ACTION_SAVE_PAGE" })
    .then(response => {
      // Handle the response from the other extension if needed
      recordLog("Received response from SingleFile extension: " + response);
    })
    .catch(error => {
      // Handle any errors that may occur
      recordError("Error sending message to SingleFile extension: " + error);
    });
```

When I execute that code, I get the message "Received response from SingleFile extension: undefined", but nothing seems to happen. Any idea of what could be failing here?

It seems that the message is correctly sent to the SingleFile extension, as if I change the targetExtensionID then I get an error "Error sending message: Error: Could not establish connection. Receiving end does not exist.". This also begs another question: the target extension ID seems to be quite arbitrary (it is not even set in the manifest.json file of SingleFile, as far as I understand), so the code above seems to be quite fragile (it depends on an ID that has no meaning for me and could change). Any idea about whether this can be managed in a better way is welcome.

Any tip would be great!

**Comment by gildas-lormeau** [comment]
The message `{ action: "ACTION_SAVE_PAGE" }` is not a valid value. You can find the valid values in the help page of the extension or here: https://github.com/gildas-lormeau/SingleFile/issues/1186#issuecomment-1646838753. Note that the message is not an object, it's just a string.

So, in you example, you have to replace `{ action: "ACTION_SAVE_PAGE" }` with `"save-page"`.

The ID of the extension is indeed arbitrary but never changes. This is a requirement in order to store data (i.e. the settings) permanently.

**Comment by silarri** [comment]
Thank you very much! Is there a way to change the default filename?

**Reply by gildas-lormeau** [reply]
You're welcome. Today, the user has to update the filename template in SingleFile to change the filename format.

**Reply by mlipok** [reply]
I try to use this:
```js
function send(extensionId, msg, callback, retry = 20) {
  const timer = setTimeout(send, 100, extensionId, msg, callback, retry - 1);
  chrome.runtime.sendMessage(extensionId, msg, response => {
    if (!chrome.runtime.lastError || !retry) {
      clearTimeout(timer);
      callback(response);
    }
  });
};
send('{531906d3-e22f-4a6c-a102-8057b88a1a63}', '{ action: "save-page" }');
```
But I get many errors.
Could you point me how to use this properly with JS ?

**Reply by gildas-lormeau** [reply]
As I have already said is this thread, `'{ action: "save-page" }'`  is not a valid value.

**Reply by urmyfaith** [reply]
two years past.

**Comment by urmyfaith** [comment]
> you have to replace { action: "ACTION_SAVE_PAGE" } with "save-page".   

eg:

```
const extensionId='mpiodijhokgodhhofbcjdecpffjipkle'
const msg='save-page';
chrome.runtime.sendMessage(extensionId, msg, response => {
    console.log(response)
});
```
